cirq/ops/thermal_relaxation_channel.py

```diff
@@ -0,0 +1,229 @@
+        and `beta` with the probability of the environment being not excited is
+        given by `p`.
+
+        The stationary state of this channel is the diagonal density matrix
+        with probability `p` of being |0⟩ and probability `1-p` of being |1⟩.
+
+        Recall the kraus opeartors for the `GeneralizedAmplitudeDampingChannel`:
+
+            $$
+            \begin{aligned}
+            M_0 =& \sqrt{p} \begin{bmatrix}
+                                1 & 0  \\
+                                0 & \sqrt{1 - \gamma}
+                            \end{bmatrix}
+            \\
+            M_1 =& \sqrt{p} \begin{bmatrix}
+                                0 & \sqrt{\gamma} \\
+                                0 & 0
+                           \end{bmatrix}
+            \\
+            M_2 =& \sqrt{1-p} \begin{bmatrix}
+                                \sqrt{1-\gamma} & 0 \\
+                                 0 & 1
+                              \end{bmatrix}
+            \\
+            M_3 =& \sqrt{1-p} \begin{bmatrix}
+                                 0 & 0 \\
+                                 \sqrt{\gamma} & 0
+                             \end{bmatrix}
+            \end{aligned}
+            $$
+
+        This has a choi representation of:
+
+            $$
+             \begin{aligned}
+                CHO =&  \begin{bmatrix}
+                            1 - (1-p) \gamma & 0 & 0 & \sqrt{1 - \gamma} \\
+                            0 & (1-p) \gamma 0 & 0 \\
+                            0 & 0 & p \gamma & 0 \\
+                            \sqrt{1 - \gamma} & 0 & 0 & 1 - p \gamma
+                        \end{bmatrix}
+             \end{aligned}
+            $$
+
+        Where each 2x2 block can be thought of as how the channel acts on each
+        basis element in the space of density matrices.
+
+        For the `ThermalRelaxationChannel` channel we simply substitute `beta`
+        in for `gamma` on the off diagonal elements of the above choi matrix
+        and then evolve the channing use the kraus operators found by sqrt
+        factorization of the choi matrix.
+
+        Args:
+            gamma: the probability of the interaction being dissipative.
+            beta: the probability of the interaction being dissipative.
+            p: the probability of the qubit and environment exchanging energy.
+
+        Raises:
+            ValueError: if gamma, beta or p is not a valid probability.
+            ValueError: if p, gamma, beta breaks CP condition.
+        """
+
+        self._gamma = value.validate_probability(gamma, 'gamma')
+        self._beta = value.validate_probability(beta, 'beta')
+        self._p = value.validate_probability(p, 'p')
+
+        choi = np.array([[1. - (1. - p) * gamma, 0, 0,
+                          np.sqrt(1. - beta)], [0, (1. - p) * gamma, 0, 0],
+                         [0, 0, p * gamma, 0],
+                         [np.sqrt(1. - beta), 0, 0, 1. - p * gamma]])
```

Comment:
We define other channels in terms of Kraus operators, not the Choi matrix. It'd be nice to follow this approach for consistency and to enable users to compare the specs of the channels.

----
It may also be a good idea to add a protocol to obtain the Choi matrix for a channel and then you could use that in a unit test to verify that the Kraus operators for `ThermalRelaxationChannel` really do yield the Choi matrix you calculated. One minor caveat in defining such a protocol is that Choi matrix is dependent on operator basis.

Reply:
Following kraus operators is pretty hard here since they aren't easy to write down. I'm also a little warry of if we need choi matrix utilities right away (since this is the first time using choi so far in the repo). Maybe if a few more cases pop up where we need it, it'd be worth making a protocol ? :)

Reply:
Sorry to barge in and argue against this PR. I believe there is a cleaner implementation that provides an easier to understand interface as well. It's based on the fact that the amplitude damping (AD) and phase damping (PD) channels commute. So really we only need two parameters to define this channel _p_AD_ and _p_PD_. The Kraus operators for the AD and PD channels can be found in equations 3 and 5 of this [paper](https://arxiv.org/pdf/1404.3747.pdf). The Kraus operators for the combined channel are just the products of the AD and PD Kraus operators; it turns out that one of the products vanishes, so you only need 3 Kraus operators.

Although you only need two parameters to define this channel, sometimes it is very convenient (from a physics perspective) to define the channel with three: evolution time _t_ , T_1, and T_2. See equation 7 for a nice interpretation of this channel with respect to these parameters. The map from these three quantities to (_p_AD_, _p_PD_) can be computed in a straightforward way (see the sentence after equation 7). So we could also make a separate factory method for this channel based on these arguments. This is nice because it ties directly to observed noise models. It also allows for straightforward merging of consecutive channels: Use the fact that the mapping is overdetermined to get the same T_1 and T_2 parameters for both channels, but different times _t_ and _t_'. Decay for time _t_ followed by decay for time _t_' is then decay for time _t_ + _t_'.

I actually already made this alternate implementation for my own use but I'm happy to share it.

Reply:
Note that currently this PR implements the composition of PD with GAD(*), not AD. These still do commute(**), but working out the minimum representation in terms of Kraus operators is a little more involved than taking the products.

That said, I do like the idea of breaking this up into simpler channels that commute.

---
(*) Generalized Amplitude Damping channel. In AD, the reservoir is at the absolute zero and consequently the Kraus operators do not include a term that excites the qubit. GAD models a reservoir with a non-zero temperature and has four Kraus operators.
(**) This is most easily seen by their action on the Bloch vector: PD shrinks rx and ry by a fixed factor and preserves rz while GAD shrinks rx and ry by a fixed factor while applying an affine function to rz.

Reply:
Regarding the question of initialization parameters, the three suggested by @dkafri make sense to me. One possible concern is the potential need for infinities to specify certain probability values, but AFAICT, these should never arise in practice.

Reply:
My mistake!

Reply:
At least in the case I considered, I believe we can avoid infinities as long as T2 <= 2 * T1, which is what is expected physically.

Reply:
I just noticed that we already have both PD and GAD implemented in cirq, so I wonder whether we really need the new channel introduced by this PR at all? @MichaelBroughton, do the existing channel work for you in combination? Perhaps all we need here is a factory function as suggested by @dkafri above?